RFC.md

```diff
@@ -1,12 +1,155 @@
+are forced to make conservative assumptions, inhibiting optimisation. 
+This information must ideally be expressible in LLVM, such that LLVM's optimisers and
+alias analysis can use this information to model multidimensional-array
+semantics.
+
+
+
+## A second, more involved example of dependence analysis going wrong
+
+```cpp
+// In an array A of size (n0 x n1), 
+// fill a subarray of size (s0 x s1)
+// which starts at an offset (o0 x o1) in the larger array A
+void set_subarray(unsigned n0, unsigned n1,
+	unsigned o0, unsigned o1,
+	unsigned s0, unsigned s1,
+	float A[n0][n1]) {
+	for (unsigned i = 0; i < s0; i++)
+		for (unsigned j = 0; j < s1; j++)
+				S: A[i + o0][j + o1] = 1;
+}
+```
+We first reduce this index expression to a sum of products:
+
+```
+(i + o0) * n0 + (j + o1) = n0i + n0o0 + j + o1
+ix(i, j, n0, n1, o0, o1) = n0i + n0o0 + j + o1
+```
+
+`ix` is the index expression which `LLVM` will see, since it is fully
+flattened, in comparison with the multi-dimensional index expression
+`index:[i + o0][j + o1] | sizes:[][n1]`.
+
+We will now show _why_ this multi-dimensional index is not always correct,
+and why guessing for one language will not work for another:
+
+Consider a call `set_subarray(n0=8, n1=9, o0=4, o1=6, s0=3, s1=6)`. At face
+value, this is incorrect if we view the array as 2D grid, since the size
+of the array is `(n0, n1) = (8, 9)`, but we are writing an array of size
+`(s0, s1) = (3, 6)` starting from `(o0, o1) = (4, 6)`. Clearly, we will
+exceed the width of the array, since `(s1 + o1 = 6 + 6 = 12) > (n1 = 9)`.
+However, now think of the array as a flattened 1D representation. In this
+case, the total size of the array is `n1xn2 = 8x9 = 72`, while the largest
+element we will access is at the largest value of `(i, j)`. That is, 
+`i = s0 - 1 = 2`, and `j = s1 - 1 = 5`.
+
+The largest index will be `ix(i=2, j=5, n0=8, n1=9, o0=4, o1=6) = 8*2+8*4+5+6=59`.
+Since `59 < 72`, we are clearly at _legal_ array indeces, by C semantics!
+
+The definition of the semantics of the language **changed the illegal
+multidimensional access** (which is illegal since it exceeds the `n1`
+dimension), into a **legal flattened 1D access** (which is legal since the
+flattened array indeces are inbounds).
+
+LLVM has no way of expressing these two different semantics. Hence, we are
+forced to:
+1. Consider flattened 1D accesses, which makes analysis of index expressions
+equivalent to analysis of polynomials over the integers, which is hard.
+2. Guess multidimensional representations, and use them at the express of
+soundness bugs as shown above.
+3. Guess multidimensional representations, use them, and check their validity
+at runtime, causing a runtime performance hit.
+
+
+Currently, Polly opts for option (3), which is to emit runtime checks, which
+will cause polly to bail out on the above program on the shown parameter
+configuration, since it detects that the multidimensional view that was
+guessed is incorrect.
```

Comment:
Would this be a good time to refer to the Optimistic Delinearization paper?

Reply:
What do you mean by "bail out"? I might imagine you mean "halt the program" or "execute an unoptimized loop" but it's probably worth being clear.

Reply:
Currently, Polly opts for option (3), which is to emit runtime checks (as described in [1]). As a result, Polly will bail out on the above program on the shown parameter configuration, since it detects that the multidimensional view that was guessed is incorrect.